Initial Work

dafuga opened the pull request.

dafuga committed:
chore: added needed dependencies

dafuga committed:
chore: added MnemonicSeed class

dafuga committed:
chore: added tests

dafuga committed:
cleanup: code cleanup

dafuga committed:
chore: updated readme

dafuga committed:
chore: using openssl-legacy-provider for node 17+

jnordberg commented on test/index.ts:
```diff
@@ -1,9 +1,36 @@
+        assert.equal(mnemonicSeed.words, words)
+    })
+
+    suite('derivePrivateKey', function () {
```
Add a test that verifies that seeds generated by a ledger maps to the public key provided by the ledger eos app

dafuga replied:
Hmm, not sure I am following. Are you saying that you want me to generate a public key using the seed (something like [this](https://github.com/iancoleman/bip39/blob/23f393acfba3f8697f42249965f4acfebae918a2/src/js/eos-util.js#L1-L7)) and compare it to the public key that is returned by core?

dafuga replied:
Just added a test for this 👍

jnordberg commented on src/index.ts:
```diff
@@ -5,9 +5,54 @@
-        n = n * p
+import {generateMnemonic, mnemonicToSeed} from 'bip39'
+import BIP32Factory from 'bip32'
+import * as ecc from 'tiny-secp256k1'
```
We should avoid pulling in an additional ecc lib if possible, core uses elliptic.js

dafuga replied:
I am trying to use `elliptic` instead, but I keep on getting a `ecc.isPoint is not a function` error. Do you know if `elliptic` has a method that checks if a point is on the elliptic curve?

jnordberg replied:
`ec('secp256k1').verify(..)` should do it I think

dafuga replied:
I don't think those methods are compatible. After mapping `isPoint` to `ec('secp256k1').verify(..)` I get a `TypeError: this._truncateToN is not a function` error.

dafuga committed:
fix: generating EOSIO compatible private keys

dafuga committed:
cleanup: code cleanup

dafuga committed:
refactor: got rid of unnecessary dependencies

dafuga committed:
enhancement: ignoring non english bip39 locales

dafuga committed:
chore: added test to make sure that library is compatible with ledger mnemonic seeds

dafuga commented:
Thanks for the feedback @jnordberg ! I added the test that you suggested. As discussed during our dev call, I think we can keep the ECC and BIP39 libraries around for now and potentially get rid of them in a future PR.

jnordberg commented on test/index.ts:
```diff
@@ -1,9 +1,44 @@
+
+    suiteSetup(function () {
+        // eslint-disable-next-line @typescript-eslint/no-var-requires
+        sinon.stub(require('bip39'), 'generateMnemonic').returns(words)
```
Why is this stubbed out? Seems like a critical thing we want to test

dafuga replied:
It's already tested in the library so I figured that we could stub it.

dafuga replied:
I guess I could remove the stub though, the tests should still work.

dafuga replied:
Cause I need a way to always obtain the same words so I can make sure that our methods are working 😄

jnordberg commented on src/index.ts:
```diff
@@ -5,9 +5,74 @@
+    }
+
+    public get hex(): Promise<string> {
+        return new Promise((resolve, reject) => {
```
Manually creating a new promise should only be done when bridging callback-style APIs to promises. This should be an async function

jnordberg commented on src/index.ts:
```diff
@@ -5,9 +5,74 @@
+            Bytes.from(bip32Interface.privateKey!)
+        )
+
+        const checksum = dsha256Checksum(privateKeyBuffer.array)
```
This isn't needed, just call the private key constructor with the data instead.

`new PrivateKey(KeyType.from('K1'), Bytes.from(privateKeyData)` (no 0x80 padding)

jnordberg reviewed (changes requested):
I'm not able to run the browser tests:

```
$ make browser-test

tests → test/browser.html...
[!] (plugin commonjs--resolver) Error: Unexpected character '' (Note that you need plugins to import files that are not JavaScript)
node_modules/tiny-secp256k1/lib/secp256k1.wasm (1:0)
generateInt32./validate_error.js``````~~~~``~<	./rand.js
   ^
2: throwError`_
3:
Error: Unexpected character '' (Note that you need plugins to import files that are not JavaScript)
    at error (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:198:30)
    at Module.error (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:12555:16)
    at Module.tryParse (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:12932:25)
    at Module.setSource (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:12837:24)
    at ModuleLoader.addModuleSource (/Users/jnordberg/Development/mnemonic-seeds/node_modules/rollup/dist/shared/rollup.js:22293:20)
```

dafuga committed:
refactor: making hex() an async function

dafuga committed:
chore: removed dsha256Checksum method

dafuga committed:
chore: added @rollup/plugin-wasm to test/rollup-config

dafuga commented:
@jnordberg thanks for the review! I added `@rollup/plugin-wasm` and I'm no longer getting that error. I also addressed some of your other feedback. Let me know if you would change anything else!

dafuga committed:
style: code formatting

dafuga commented:
@jnordberg Any thought on this? 😄

dafuga commented:
I'm gonna go ahead and merge this since it's functional. If we have ideas on how to make it better we can open a new PR.
